### PR TITLE
CSS Inlining

### PR DESCRIPTION
**Pull request by developit**
This adds a new plugin (which I think could be released on its own, and I have a funny name for). It plugs into the output of `html-webpack-plugin`, finds inline & external stylesheets, inlines them, then trims out selectors & rules not used by the document.

In this app, the effect is a drop from 13kb to 1.2kb of critical CSS.

(note: currently targeted at `app-skeleton` branch)

Still to do:

- [x] If the stylesheet being inlined is external, replace the `<link>` tag with an async one (preload -> style).

/cc @surma

### CODE REVIEW

File: config/html-webpack-inline-critical-css-plugin.js [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+
+const PLUGIN_NAME = 'html-webpack-inline-critical-css-plugin';
+
+const PARSE5_OPTS = {
+  treeAdapter: treeUtils
+};
+
+function defineProperties(obj, properties) {
+  for (let i in properties) {
+    let value = properties[i];
```
Comment (mathiasbynens) [inline config/html-webpack-inline-critical-css-plugin.js]: Why are these not `const`?

File: config/html-webpack-inline-critical-css-plugin.js [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+  nodeType: {
+    get: function () {
+      return 11;
+    }
+  },
+  createElement: function (name) {
+    return treeUtils.createElement(name, null, []);
+  },
+  createTextNode: function (text) {
+    let scratch = this.$$scratchElement;
```
Comment (mathiasbynens) [inline config/html-webpack-inline-critical-css-plugin.js]: `const`

File: config/html-webpack-inline-critical-css-plugin.js [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+    treeUtils.insertText(scratch, text);
+    let node = scratch.lastChild;
```
Comment (mathiasbynens) [inline config/html-webpack-inline-critical-css-plugin.js]: Same here

Comment (mathiasbynens) [reply config/html-webpack-inline-critical-css-plugin.js]: …and throughout the rest of the file, really

File: config/html-webpack-inline-critical-css-plugin.js (hunk 1 above)
Comment (jakearchibald) [reply config/html-webpack-inline-critical-css-plugin.js]: +1 we should only use `let` if the value is changed.

File: config/html-webpack-inline-critical-css-plugin.js [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+        let externalSheets;
+        externalSheets = document.querySelectorAll('link[rel="stylesheet"]');
+
+        Promise.all(externalSheets.map(function(link) {
+          const href = link.getAttribute('href');
+          if (href.match(/^(https?:)?\/\//)) return Promise.resolve();
+          const filename = path.resolve(outputPath, href.replace(/^\//, ''));
+          let asset = compilation.assets[path.relative(outputPath, filename).replace(/^\.\//, '')];
+          let promise = asset ? Promise.resolve(asset.source()) : readFile(filename);
+          return promise.then(function (sheet) {
```
Comment (jakearchibald) [inline config/html-webpack-inline-critical-css-plugin.js]: Would this be easier to read with async functions?

File: config/html-webpack-inline-critical-css-plugin.js (hunk 1 above)
Comment (developit) [reply config/html-webpack-inline-critical-css-plugin.js]: whoop, old habits die hard. I imagine there's a lint rule to remind me of this.

File: config/html-webpack-inline-critical-css-plugin.js (hunk 4 above)
Comment (developit) [reply config/html-webpack-inline-critical-css-plugin.js]: yup. I was avoiding the transpile step. that said, I would vastly prefer async here.

**Comment by developit** [comment]
@jakearchibald Might be worth adding an option to scan for strings that look like classnames in inline JS?  Or even simpler, an option to force the existence of selector parts.

Right now this is "css required to style the static html".

File: config/html-webpack-inline-critical-css-plugin.js [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+    if (!sheet) return done;
+
+    const ast = css.parse(sheet);
+
+    const before = sheet;
+
+    visit(ast, function (rule) {
+      if (rule.type==='rule') {
+        rule.selectors = rule.selectors.filter(function (sel) {
+          if (sel.match(/:(hover|focus|active)([.[#~&^:*]|\s|\n|$)/)) return false;
```
Comment (jakearchibald) [inline config/html-webpack-inline-critical-css-plugin.js]: Does this mean `:hover` rules etc are always lost?

File: config/html-webpack-inline-critical-css-plugin.js [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+
+const treeUtils = parse5.treeAdapters.htmlparser2;
+
+const PLUGIN_NAME = 'html-webpack-inline-critical-css-plugin';
+
+const PARSE5_OPTS = {
+  treeAdapter: treeUtils
+};
+
+function defineProperties(obj, properties) {
```
Comment (surma) [inline config/html-webpack-inline-critical-css-plugin.js]: Why are you polyfilling `defineProperties`?

File: config/html-webpack-inline-critical-css-plugin.js [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+  }
+}
+
+const ElementExtensions = {
+  nodeName: {
+    get: function() {
+      return this.tagName;
+    }
+  },
+  appendChild: function (child) {
```
Comment (surma) [inline config/html-webpack-inline-critical-css-plugin.js]: I can haz `appendChild(child) { /*...*/ }` syntax?

File: config/html-webpack-inline-critical-css-plugin.js [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+
+  apply(compiler) {
+    const self = this;
+    const outputPath = compiler.options.output.path;
+    compiler.hooks.compilation.tap(PLUGIN_NAME, function (compilation) {
+      compilation.hooks.htmlWebpackPluginAfterHtmlProcessing.tapAsync(PLUGIN_NAME, function (htmlPluginData, callback) {
+        const document = parse5.parse(htmlPluginData.html, PARSE5_OPTS);
+
+        defineProperties(document, DocumentExtensions);
+        document.documentElement = document.childNodes[document.childNodes.length - 1];
```
Comment (surma) [inline config/html-webpack-inline-critical-css-plugin.js]: ... Is that guaranteed to be correct?

File: config/html-webpack-inline-critical-css-plugin.js [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+          CACHING: false,
+          USE_QSAPI: false,
+          USE_HTML5: false
+        });
+
+        const externalSheets = document.querySelectorAll('link[rel="stylesheet"]');
+
+        Promise.all(externalSheets.map(function(link) {
+          const href = link.getAttribute('href');
+          if (href.match(/^(https?:)?\/\//)) return Promise.resolve();
```
Comment (surma) [inline config/html-webpack-inline-critical-css-plugin.js]: This any stylesheet that is included via a full URL is ignored, right?

File: config/html-webpack-inline-critical-css-plugin.js [hunk 10]
```diff
@@ -0,0 +1,255 @@
...
+  // https://github.com/dperini/nwmatcher/blob/3edb471e12ce7f7d46dc1606c7f659ff45675a29/src/nwmatcher.js#L353
+  addEventListener: Object
+};
+
+module.exports = class HtmlWebpackInlineCriticalCssPlugin {
+  constructor(options) {
+    this.options = options || {};
+  }
+
+  apply(compiler) {
```
Comment (surma) [inline config/html-webpack-inline-critical-css-plugin.js]: Willing to make this async and use `async`/`await`?

File: config/html-webpack-inline-critical-css-plugin.js (hunk 5 above)
Comment (developit) [reply config/html-webpack-inline-critical-css-plugin.js]: yes. I was waffling on them and I think I might remove this clause - hover and focus are valid states for a static HTML document.

File: config/html-webpack-inline-critical-css-plugin.js [hunk 11]
```diff
@@ -0,0 +1,255 @@
...
+        document.$match = nwmatcher({ document });
+        document.$match.configure({
+          CACHING: false,
+          USE_QSAPI: false,
+          USE_HTML5: false
+        });
+
+        const externalSheets = document.querySelectorAll('link[rel="stylesheet"]');
+
+        Promise.all(externalSheets.map(function(link) {
```
Comment (surma) [inline config/html-webpack-inline-critical-css-plugin.js]: Maybe put the map function in a named top-level function so that the name describes what its doing?

File: config/html-webpack-inline-critical-css-plugin.js [hunk 12]
```diff
@@ -0,0 +1,255 @@
...
+            callback(null, { html });
+          })
+          .catch(function (err) {
+            callback(err);
+          });
+      });
+    });
+  }
+
+  processStyle(style, document) {
```
Comment (surma) [inline config/html-webpack-inline-critical-css-plugin.js]: Comment?

File: config/html-webpack-inline-critical-css-plugin.js [hunk 13]
```diff
@@ -0,0 +1,255 @@
...
+          const href = link.getAttribute('href');
+          if (href.match(/^(https?:)?\/\//)) return Promise.resolve();
+          const filename = path.resolve(outputPath, href.replace(/^\//, ''));
+          const asset = compilation.assets[path.relative(outputPath, filename).replace(/^\.\//, '')];
+          const promise = asset ? Promise.resolve(asset.source()) : readFile(filename);
+          return promise.then(function (sheet) {
+            const style = document.createElement('style');
+            style.$$name = href;
+            style.appendChild(document.createTextNode(sheet));
+            link.parentNode.appendChild(style);  // @TODO insertBefore
```
Comment (surma) [inline config/html-webpack-inline-critical-css-plugin.js]: If you `return style` here, then you don’t need to `byTag()` in the next `.then()`, no?

File: config/html-webpack-inline-critical-css-plugin.js (hunk 5 above)
Comment (surma) [reply config/html-webpack-inline-critical-css-plugin.js]: I thought this meant that they are always kept, but I am also not 100% sure. Comments please? 😄

File: config/html-webpack-inline-critical-css-plugin.js [hunk 14]
```diff
@@ -0,0 +1,255 @@
...
+        // If there are no remaining rules, replace the parent rule with a comment.
+        if (rule.rules.length===0) {
+          rule.type = 'comment';
+          rule.comment = '';
+          delete rule.rules;
+        }
+      }
+    });
+
+    sheet = css.stringify(ast, { compress: true });
```
Comment (surma) [inline config/html-webpack-inline-critical-css-plugin.js]: `... {compress: this.options.compress}`?

File: config/html-webpack-inline-critical-css-plugin.js [hunk 15]
```diff
@@ -0,0 +1,255 @@
...
+
+    if (this.options.minimize || this.options.compress || this.options.minify) {
+      const cssnano = require('cssnano');
```
Comment (surma) [inline config/html-webpack-inline-critical-css-plugin.js]: Move this to the top.

File: config/html-webpack-inline-critical-css-plugin.js [hunk 16]
```diff
@@ -0,0 +1,255 @@
...
+  processStyle(style, document) {
+    let done = Promise.resolve();
+    let sheet = style.childNodes.length>0 && style.childNodes.map(getNodeValue).join('\n');
+    if (!sheet) return done;
+
+    const ast = css.parse(sheet);
+
+    const before = sheet;
+
+    visit(ast, function (rule) {
```
Comment (surma) [inline config/html-webpack-inline-critical-css-plugin.js]: The callback could be async as well

File: config/html-webpack-inline-critical-css-plugin.js [hunk 17]
```diff
@@ -0,0 +1,255 @@
...
+  node.rules.forEach(function (rule) {
+    if (rule.rules) {
+      visit(rule, fn);
+    }
+    fn(rule);
+  });
+}
+
+
+function readFile(file) {
```
Comment (surma) [inline config/html-webpack-inline-critical-css-plugin.js]: WDYT about:

```js
const {promisify} = require('utils');
const fs = require('fs');
const readFile = promisify(fs.readFile);
```

File: config/html-webpack-inline-critical-css-plugin.js [hunk 18]
```diff
@@ -0,0 +1,255 @@
...
+
+function getNodeValue(node) {
+  return node.nodeValue;
+}
+
+function notComment(rule) {
+  return rule.type !== 'comment';
+}
+
+function getElementsByTagName(tagName) {
```
Comment (surma) [inline config/html-webpack-inline-critical-css-plugin.js]: I mean, it doesn’t really matter. But a recursive function could be easier to read:

```js
function getElementsByTagName(tagName) {
  if (this.nodeType !== 1) return []; // Only elements are relevant
  tagName = tagName.toUpperCase();
  return [
    ...((this.tagName == tagName || tagName === '*') ? tagName : [])), // Add current element if it matches tag
    ...this.children.map(child => getElementByTagName(tagName).call(child)) // Check children recursively
  ];
}
```

File: config/html-webpack-inline-critical-css-plugin.js [hunk 19]
```diff
@@ -0,0 +1,255 @@
...
+};
+
+function defineProperties(obj, properties) {
+  for (const i in properties) {
+    const value = properties[i];
+    Object.defineProperty(obj, i, typeof value === 'function' ? { value: value } : value);
+  }
+}
+
+const ElementExtensions = {
```
Comment (surma) [inline config/html-webpack-inline-critical-css-plugin.js]: Comment?

File: config/html-webpack-inline-critical-css-plugin.js [hunk 20]
```diff
@@ -0,0 +1,255 @@
...
+    return this.attribs != null && this.attribs[name] != null;
+  },
+  getAttributeNode: function (name) {
+    const value = this.getAttribute(name);
+    if (value!=null) return { specified: true, value: value };
+  },
+  getElementsByTagName: getElementsByTagName
+};
+
+const DocumentExtensions = {
```
Comment (surma) [inline config/html-webpack-inline-critical-css-plugin.js]: Comment?

**Comment by surma** [comment]
Oh and I forgot: Can we have some tests infra for this? I’d love to use this, but I’d like to be able to test (esp. if you wanna go ahead and publish this and we start running into corner cases).

File: config/html-webpack-inline-critical-css-plugin.js (hunk 6 above)
Comment (developit) [reply config/html-webpack-inline-critical-css-plugin.js]: Wasn't - most of the values are functions that get assigned as pure `value`s, this made that a bit easier (rather than having to specific `{ value() {} }` for each prototype method.

File: config/html-webpack-inline-critical-css-plugin.js (hunk 7 above)
Comment (developit) [reply config/html-webpack-inline-critical-css-plugin.js]: Just looked and its Node 4+ so yes!

File: config/html-webpack-inline-critical-css-plugin.js (hunk 6 above)
Comment (surma) [reply config/html-webpack-inline-critical-css-plugin.js]: Oh I see. Comment please :D

File: config/html-webpack-inline-critical-css-plugin.js [hunk 21]
```diff
@@ -0,0 +1,255 @@
...
+        document.$match.configure({
+          CACHING: false,
+          USE_QSAPI: false,
+          USE_HTML5: false
+        });
+
+        const externalSheets = document.querySelectorAll('link[rel="stylesheet"]');
+
+        Promise.all(externalSheets.map(function(link) {
+          const href = link.getAttribute('href');
```
Comment (mathiasbynens) [inline config/html-webpack-inline-critical-css-plugin.js]: Use `link.href` to always get a normalized URL string (even when the attribute value is a relative URL, for example).

File: config/html-webpack-inline-critical-css-plugin.js (hunk 7 above)
Comment (surma) [reply config/html-webpack-inline-critical-css-plugin.js]: Out of curiosity — what is your baseline for build tooling like this? My gut feeling would be LTS, so Node 8.

Comment (surma) [reply config/html-webpack-inline-critical-css-plugin.js]: Nvm, asked this in chat.

File: config/html-webpack-inline-critical-css-plugin.js (hunk 21 above)
Comment (developit) [reply config/html-webpack-inline-critical-css-plugin.js]: nah, this DOM (parse5) is extremely rudimentary and only supports attributes.

File: config/html-webpack-inline-critical-css-plugin.js (hunk 8 above)
Comment (developit) [reply config/html-webpack-inline-critical-css-plugin.js]: Nope. I'll have it filter by tagName.

File: config/html-webpack-inline-critical-css-plugin.js (hunk 9 above)
Comment (developit) [reply config/html-webpack-inline-critical-css-plugin.js]: Yes. I might also add an attribute option to filter by, or make this filterable via a regex passed as an option.

File: config/html-webpack-inline-critical-css-plugin.js (hunk 11 above)
Comment (developit) [reply config/html-webpack-inline-critical-css-plugin.js]: Just annoying to bind.

File: config/html-webpack-inline-critical-css-plugin.js (hunk 13 above)
Comment (developit) [reply config/html-webpack-inline-critical-css-plugin.js]: It scans for all style tags, not just those embedded by this function.

File: config/html-webpack-inline-critical-css-plugin.js (hunk 14 above)
Comment (developit) [reply config/html-webpack-inline-critical-css-plugin.js]: I need to compress it in order to have it remove the comments (that's how I'm doing AST node removal).

File: config/html-webpack-inline-critical-css-plugin.js (hunk 15 above)
Comment (developit) [reply config/html-webpack-inline-critical-css-plugin.js]: But then it'll be imported regardless of whether it gets used

**Comment by surma** [comment]
I’m gonna wait with my 2nd review until you push the async/await :)

File: config/html-webpack-inline-critical-css-plugin.js (hunk 10 above)
Comment (developit) [reply config/html-webpack-inline-critical-css-plugin.js]: yup

File: config/html-webpack-inline-critical-css-plugin.js (hunk 14 above)
Comment (developit) [reply config/html-webpack-inline-critical-css-plugin.js]: I've removed the cssnano integration - never resulted in any compression for me.

File: config/html-webpack-inline-critical-css-plugin.js (hunk 15 above)
Comment (developit) [reply config/html-webpack-inline-critical-css-plugin.js]: remove this entirely

File: config/html-webpack-inline-critical-css-plugin.js (hunk 16 above)
Comment (developit) [reply config/html-webpack-inline-critical-css-plugin.js]: how come? it's a synchronous walk.

File: config/html-webpack-inline-critical-css-plugin.js (hunk 17 above)
Comment (developit) [reply config/html-webpack-inline-critical-css-plugin.js]: yup, if we're doing 8+ that's all this does.

File: config/html-webpack-inline-critical-css-plugin.js (hunk 18 above)
Comment (developit) [reply config/html-webpack-inline-critical-css-plugin.js]: easier to read but quite a bit slower. also the weird edge cases are to handle the fact that this dom doesn't normalize the case of tagName

Comment (developit) [reply config/html-webpack-inline-critical-css-plugin.js]: Whoop, here I was spouting "facts" without testing. Recursion is *much* faster here:

<a href="https://esbench.com/bench/5ac3b647f2949800a0f619e1"></a>

I'll update with the ones from that [bench](https://esbench.com/bench/5ac3b647f2949800a0f619e1).

Comment (surma) [reply config/html-webpack-inline-critical-css-plugin.js]: Even if it wasn’t, I’d prefer the more readable/concise version (unless the difference is on the magnitude of seconds). It’s a build step after all.

Comment (surma) [reply config/html-webpack-inline-critical-css-plugin.js]: Also we can probably safe the `bind()` and turn the element into a normal parameter, if that makes it more readable.

Comment (developit) [reply config/html-webpack-inline-critical-css-plugin.js]: The bind is because this gets mixed into Element.prototype.

My benchmark was flawed, stack is around 5x faster, or 2x faster than an optimized recursive version:



(same url: https://esbench.com/bench/5ac3b647f2949800a0f619e1)

File: config/critters-webpack-plugin.js [hunk 22]
```diff
@@ -0,0 +1,304 @@
...
+
+        // `external:false` skips processing of external sheets
+        if (this.options.external!==false) {
+          const externalSheets = document.querySelectorAll('link[rel="stylesheet"]');
+          externalStylesProcessed = Promise.all(externalSheets.map(
+            link => this.embedLinkedStylesheet(link, compilation, outputPath)
+          ));
+        }
+
+        externalStylesProcessed
```
Comment (surma) [inline config/critters-webpack-plugin.js]: async/await?

File: config/critters-webpack-plugin.js [hunk 23]
```diff
@@ -0,0 +1,304 @@
...
+    return fn(rule)!==false;
+  });
+}
+
+
+/** Enhance an htmlparser2-style DOM with basic manipulation methods. */
+function makeDomInteractive(document) {
+  defineProperties(document, DocumentExtensions);
+  // Find the first <html> element within the document
+  document.documentElement = document.childNodes.filter( child => String(child.tagName).toLowerCase()==='html' )[0];
```
Comment (surma) [inline config/critters-webpack-plugin.js]: Have you linted this file with semistandard? The spaces look suspicious here

File: config/critters-webpack-plugin.js [hunk 24]
```diff
@@ -0,0 +1,304 @@
...
+
+    // path on disk
+    const filename = path.resolve(outputPath, href.replace(/^\//, ''));
+
+    // try to find a matching asset by filename in webpack's output (not yet written to disk)
+    const asset = compilation.assets[path.relative(outputPath, filename).replace(/^\.\//, '')];
+
+    // wait for a disk read if we had to go to disk
+    const promise = asset ? Promise.resolve(asset.source()) : readFile(filename, 'utf8');
+    return promise.then(sheet => {
```
Comment (surma) [inline config/critters-webpack-plugin.js]: async/await?

File: config/critters-webpack-plugin.js [hunk 25]
```diff
@@ -0,0 +1,304 @@
...
+      link.parentNode.insertBefore(style, link.nextSibling);
+
+      // drop a reference to the original URL onto the tag (used for reporting to console later)
+      style.$$name = href;
+
+      // the `async` option changes any critical'd <link rel="stylesheet"> tags to async-loaded equivalents
+      if (this.options.async) {
+        link.setAttribute('rel', 'preload');
+        link.setAttribute('as', 'style');
+        link.setAttribute('onload', "this.rel='stylesheet'");
```
Comment (surma) [inline config/critters-webpack-plugin.js]: What does this `onload` handler do o_O

File: config/critters-webpack-plugin.js [hunk 26]
```diff
@@ -0,0 +1,304 @@
...
+        }
+      }
+
+      // If there are no remaining rules, remove the whole rule.
+      return !rule.rules || rule.rules.length!==0;
+    });
+
+    sheet = css.stringify(ast, { compress: this.options.compress!==false });
+
+    return done.then(() => {
```
Comment (surma) [inline config/critters-webpack-plugin.js]: async/await? 😬

**Comment by surma** [comment]
Wait... did I just put “async/await?” everywhere and you are still working on that? Apologies if that’s the case. I thought you pushed a commit. Sorryyy

File: config/critters-webpack-plugin.js (hunk 22 above)
Comment (developit) [reply config/critters-webpack-plugin.js]: coming :)

File: config/critters-webpack-plugin.js (hunk 23 above)
Comment (developit) [reply config/critters-webpack-plugin.js]: I'll switch over - right now it is using my fork but needn't be.

File: config/critters-webpack-plugin.js (hunk 25 above)
Comment (developit) [reply config/critters-webpack-plugin.js]: swaps the preload to a stylesheet once it's loaded:
https://www.filamentgroup.com/lab/async-css.html

**Comment by developit** [comment]
lol still workin on it. will be up in a jiffy - my day got torched yesterday.

File: config/html-webpack-inline-critical-css-plugin.js (hunk 21 above)
Comment (mathiasbynens) [reply config/html-webpack-inline-critical-css-plugin.js]: Ooh, gotcha. Fingers crossed dealing with raw attribute values doesn’t break anything, I guess!
